### PR TITLE
Docs and Improvements

### PR DESCRIPTION
**Pull request by Shqrdx**
New docs and improvements !

<h3>Description</h3>

<h4>Quaternion</h4> 
 -Docs for all things
 -New constructor which accepts single Vector4
 -DebugDisplayString replaced to correct place
-Concatenate,CreateFromAxisAngle,CreateFromYawPitchRoll,CreateFromYawPitchRoll,Conjugate,        Equals,Normalize.. rewrited to more nice view(with new tests of course).
 -Redundant "this" and (double) conversion removed from some methods
 -Removed unused internal methods

<h4>All vectors</h4> 
  changed few docs

<h4>Vector4</h4>
 <strike> -New constructor which accepts Quaternion</strike>
 -Added ToQuaternion method

<h4>Tests</h4>
-All vector tests rewrited to more compact view - added new Compare method which is usable for comparsion of float-point structs without redundant code. 
<strike>-New file for test called FPointBase -provide Compare method for comparsion between 2 floats.</strike>
-Added new QuaternionTest which follows the rules above

### CODE REVIEW

**Comment by Shqrdx** [comment]
@tomspilman @KonajuGames - your thoughts ?

**Comment by tomspilman** [comment]
Haven't had a chance to go thru it yet... I will soon.

**Comment by Shqrdx** [comment]
4(7) days passed and I still waiting :)

File: MonoGame.Framework/Quaternion.cs [hunk 1]
```diff
@@ -8,25 +8,56 @@
 
 namespace Microsoft.Xna.Framework
 {
+    /// <summary>
+    /// Describes a 4D-vector which designed for rotation operations.
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```An efficient mathematical representation for three dimensional rotations.```

File: MonoGame.Framework/Quaternion.cs [hunk 2]
```diff
@@ -34,25 +65,80 @@ public Quaternion(float x, float y, float z, float w)
...
+            this.Z = value.Z;
+            this.W = value.W;
         }
 
+        #endregion
+
+        #region Public Properties
+
+        /// <summary>
+        /// Returns a <see cref="Quaternion"/> with components 0, 0, 0, 1(without the rotation).
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```Returns a quaternion representing no rotation.```

File: MonoGame.Framework/Quaternion.cs [hunk 3]
```diff
@@ -61,242 +147,331 @@ public static Quaternion Add(Quaternion quaternion1, Quaternion quaternion2)
...
+
+        #endregion
+
+        #region CreateFromAxisAngle
+
+        /// <summary>
+        /// Creates a new <see cref="Quaternion"/> from the specified axis and angle.
+        /// </summary>
+        /// <param name="axis">The axis of rotation.</param>
+        /// <param name="angle">The angle to rotate around the axis.</param>
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```The angle in radians.```

File: MonoGame.Framework/Quaternion.cs [hunk 4]
```diff
@@ -61,242 +147,331 @@ public static Quaternion Add(Quaternion quaternion1, Quaternion quaternion2)
...
+		    float cos = (float)Math.Cos(half);
+		    return new Quaternion(axis.X * sin, axis.Y * sin, axis.Z * sin, cos);
         }
 
-
+        /// <summary>
+        /// Creates a new <see cref="Quaternion"/> from the specified axis and angle.
+        /// </summary>
+        /// <param name="axis">The axis of rotation.</param>
+        /// <param name="angle">The angle to rotate around the axis.</param>
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```The angle in radians.```

File: MonoGame.Framework/Quaternion.cs [hunk 5]
```diff
@@ -61,242 +147,331 @@ public static Quaternion Add(Quaternion quaternion1, Quaternion quaternion2)
...
         }
 
+        #endregion
+
+        #region CreateFromRotationMatrix
 
+        /// <summary>
+        /// Creates a new <see cref="Quaternion"/> from the specified <see cref="Matrix"/>.
+        /// </summary>
+        /// <param name="matrix">The matrix which contains rotation information.</param>
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```The rotation matrix.```

File: MonoGame.Framework/Quaternion.cs [hunk 6]
```diff
@@ -61,242 +147,331 @@ public static Quaternion Add(Quaternion quaternion1, Quaternion quaternion2)
...
+        /// <returns>The new quaternion builded from matrix.</returns>
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```A quaternion composed from the rotation part of the matrix.```

File: MonoGame.Framework/Quaternion.cs [hunk 7]
```diff
@@ -61,242 +147,331 @@ public static Quaternion Add(Quaternion quaternion1, Quaternion quaternion2)
...
 		    return quaternion;
-
         }
 
-
+        /// <summary>
+        /// Creates a new <see cref="Quaternion"/> from the specified <see cref="Matrix"/>.
+        /// </summary>
+        /// <param name="matrix">The matrix which contains rotation information.</param>
+        /// <param name="result">The new quaternion builded from matrix as an output parameter.</param>
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```A quaternion composed from the rotation part of the matrix.```

File: MonoGame.Framework/Quaternion.cs [hunk 8]
```diff
@@ -61,242 +147,331 @@ public static Quaternion Add(Quaternion quaternion1, Quaternion quaternion2)
...
 			
 		    return quaternion;
-
         }
 
-
+        /// <summary>
+        /// Creates a new <see cref="Quaternion"/> from the specified <see cref="Matrix"/>.
+        /// </summary>
+        /// <param name="matrix">The matrix which contains rotation information.</param>
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```The rotation matrix.```

File: MonoGame.Framework/Quaternion.cs [hunk 9]
```diff
@@ -61,242 +147,331 @@ public static Quaternion Add(Quaternion quaternion1, Quaternion quaternion2)
...
+        #endregion
+
+        #region CreateFromYawPitchRoll
+
+        /// <summary>
+        /// Creates a new <see cref="Quaternion"/> from the specified yaw, pitch and roll scalars.
+        /// </summary>
+        /// <param name="yaw"></param>
+        /// <param name="pitch"></param>
+        /// <param name="roll"></param>
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```Roll around the z axis in radians.```

File: MonoGame.Framework/Quaternion.cs [hunk 10]
```diff
@@ -61,242 +147,331 @@ public static Quaternion Add(Quaternion quaternion1, Quaternion quaternion2)
...
+
+        #endregion
+
+        #region CreateFromYawPitchRoll
+
+        /// <summary>
+        /// Creates a new <see cref="Quaternion"/> from the specified yaw, pitch and roll scalars.
+        /// </summary>
+        /// <param name="yaw"></param>
+        /// <param name="pitch"></param>
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```Pitch around the x axis in radians.```

File: MonoGame.Framework/Quaternion.cs [hunk 11]
```diff
@@ -61,242 +147,331 @@ public static Quaternion Add(Quaternion quaternion1, Quaternion quaternion2)
...
-		public static Quaternion CreateFromYawPitchRoll(float yaw, float pitch, float roll)
+
+        #endregion
+
+        #region CreateFromYawPitchRoll
+
+        /// <summary>
+        /// Creates a new <see cref="Quaternion"/> from the specified yaw, pitch and roll scalars.
+        /// </summary>
+        /// <param name="yaw"></param>
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```Yaw around the y axis in radians.```

**Comment by Shqrdx** [comment]
I have a question about how to use "The" - I need to rewrite it to "A" in (for example) "The new quaternion builded from the specified yaw, pitch and roll scalars."

File: MonoGame.Framework/Quaternion.cs [hunk 12]
```diff
@@ -61,242 +147,331 @@ public static Quaternion Add(Quaternion quaternion1, Quaternion quaternion2)
...
+        /// <param name="pitch"></param>
+        /// <param name="roll"></param>
+        /// <returns>The new quaternion builded from the specified yaw, pitch and roll scalars.</returns>
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```A new quaternion from the concatenated yaw, pitch, and roll angles.```

Comment (Shqrdx) [inline MonoGame.Framework/Quaternion.cs]: Creates a new <see cref="Quaternion"/> from the specified yaw, pitch and roll scalars. - should I change it too ?

Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: > Creates a new from the specified yaw, pitch and roll scalars. - should I change it too ?

I don't understand what you are referring to.  What I wrote above is correct.

**Comment by Shqrdx** [comment]
Damn, this bot takes half-hour to check a simple change..

File: MonoGame.Framework/Quaternion.cs (hunk 12 above)
Comment (Shqrdx) [inline MonoGame.Framework/Quaternion.cs]: I mean the doc at line 414

Comment (Shqrdx) [inline MonoGame.Framework/Quaternion.cs]: I think you write it for doc at 419.

File: MonoGame.Framework/Quaternion.cs [hunk 13]
```diff
@@ -342,45 +522,76 @@ public static void Divide(ref Quaternion quaternion1, ref Quaternion quaternion2
...
+        /// <returns><c>true</c> if the instances are equal; <c>false</c> otherwise.</returns>
         public override bool Equals(object obj)
         {
-             bool flag = false;
-		    if (obj is Quaternion)
-		    {
-		        flag = this.Equals((Quaternion) obj);
-		    }
-		    return flag;
+            return (obj is Quaternion) ? this == (Quaternion)obj : false;
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: If you are going to change this we might as well make the code clear and simple.
```
if (obj is Quaternion)
   return Equals((Quaternion)obj);
return false;
```

File: MonoGame.Framework/Quaternion.cs (hunk 12 above)
Comment (Shqrdx) [inline MonoGame.Framework/Quaternion.cs]: I think change - "Creates a new quaternion from the specified yaw, pitch and roll angles", looks better.

Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: Sure... that is fine.

File: MonoGame.Framework/Quaternion.cs [hunk 14]
```diff
@@ -342,45 +522,76 @@ public static void Divide(ref Quaternion quaternion1, ref Quaternion quaternion2
...
         public override int GetHashCode()
         {
-            return (((this.X.GetHashCode() + this.Y.GetHashCode()) + this.Z.GetHashCode()) + this.W.GetHashCode());
+            return X.GetHashCode() + Y.GetHashCode() + Z.GetHashCode() + W.GetHashCode();
         }
 
+        #region Inverse
 
+        /// <summary>
+        /// Creates a new <see cref="Quaternion"/> that contains an inversion of specified quaternion.
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```Returns the inverse quaternion which represents the opposite rotation.```

**Comment by Shqrdx** [comment]
@tomspilman anything else ?

File: MonoGame.Framework/Quaternion.cs [hunk 15]
```diff
@@ -404,19 +619,36 @@ public static void Inverse(ref Quaternion quaternion, out Quaternion result)
 		    result.W = quaternion.W * num;
         }
 
+        #endregion
+
+        /// <summary>
+        /// Returns the length of this <see cref="Quaternion"/>.
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```Returns the magnitude of the quaternion components.```

File: MonoGame.Framework/Quaternion.cs [hunk 16]
```diff
@@ -404,19 +619,36 @@ public static void Inverse(ref Quaternion quaternion, out Quaternion result)
...
-    		return (float) Math.Sqrt((double) num);
+            float num = (X * X) + (Y * Y) + (Z * Z) + (W * W);
+    		return (float) Math.Sqrt(num);
         }
 
-
+        /// <summary>
+        /// Returns the squared length of this <see cref="Quaternion"/>.
+        /// </summary>
+        /// <returns>The squared length of this <see cref="Quaternion"/>.</returns>
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```Returns the squared magnitude of the quaternion components.```

File: MonoGame.Framework/Quaternion.cs [hunk 17]
```diff
@@ -404,19 +619,36 @@ public static void Inverse(ref Quaternion quaternion, out Quaternion result)
...
         public float LengthSquared()
         {
-            return ((((this.X * this.X) + (this.Y * this.Y)) + (this.Z * this.Z)) + (this.W * this.W));
+            return (X * X) + (Y * Y) + (Z * Z) + (W * W);
         }
 
+        #region Lerp
 
+        /// <summary>
+        /// Creates a new <see cref="Quaternion"/> that contains linear interpolation of the specified quaternions.
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```Performs a linear blend between two quaternions.```

File: MonoGame.Framework/Quaternion.cs [hunk 18]
```diff
@@ -404,19 +619,36 @@ public static void Inverse(ref Quaternion quaternion, out Quaternion result)
...
+        /// </summary>
+        /// <param name="quaternion1">Source <see cref="Vector4"/>.</param>
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: Vector4?

Comment (Shqrdx) [inline MonoGame.Framework/Quaternion.cs]: error.. quaternion of course.

File: MonoGame.Framework/Quaternion.cs [hunk 19]
```diff
@@ -404,19 +619,36 @@ public static void Inverse(ref Quaternion quaternion, out Quaternion result)
...
+        /// <param name="quaternion2">Source <see cref="Vector4"/>.</param>
+        /// <param name="amount">Weighting value.</param>
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```The blend amount where 0 returns the first quaternion and 1 returns the second quaternion.```

File: MonoGame.Framework/Quaternion.cs [hunk 20]
```diff
@@ -475,7 +713,17 @@ public static void Lerp(ref Quaternion quaternion1, ref Quaternion quaternion2,
 
         }
 
+        #endregion
+
+        #region Slerp
 
+        /// <summary>
+        /// Creates a new <see cref="Quaternion"/> that contains spherical linear interpolation of the specified quaternions.
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```Performs a spherical linear blend between two quaternions.```

File: MonoGame.Framework/Quaternion.cs [hunk 21]
```diff
@@ -475,7 +713,17 @@ public static void Lerp(ref Quaternion quaternion1, ref Quaternion quaternion2,
...
+        /// </summary>
+        /// <param name="quaternion1">Source <see cref="Vector4"/>.</param>
+        /// <param name="quaternion2">Source <see cref="Vector4"/>.</param>
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: Vector4?

Comment (Shqrdx) [inline MonoGame.Framework/Quaternion.cs]: quaternion.. I know..

File: MonoGame.Framework/Quaternion.cs [hunk 22]
```diff
@@ -623,18 +915,25 @@ public static void Multiply(ref Quaternion quaternion1, ref Quaternion quaternio
 		    result.W = (w * num) - num9;
         }
 
+        #endregion
+
+        #region Negate
 
+        /// <summary>
+        /// Creates a new <see cref="Quaternion"/> that contains the specified quaternion inversion.
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```Flips the sign of the all the quaternion components.```

File: MonoGame.Framework/Quaternion.cs [hunk 23]
```diff
@@ -623,18 +915,25 @@ public static void Multiply(ref Quaternion quaternion1, ref Quaternion quaternio
...
+        /// </summary>
+        /// <param name="quaternion">Source <see cref="Quaternion"/>.</param>
+        /// <returns>The result of the quaternion inversion.</returns>
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```Returns the negated quaternion.```

File: MonoGame.Framework/Quaternion.cs [hunk 24]
```diff
@@ -643,42 +942,83 @@ public static void Negate(ref Quaternion quaternion, out Quaternion result)
 		    result.W = -quaternion.W;
         }
 
+        #endregion
+
+        #region Normalize
 
+        /// <summary>
+        /// Turns this <see cref="Quaternion"/> to a unit quaternion with the same direction.
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```Scales the quaternion magnitude to unit length.```

File: MonoGame.Framework/Quaternion.cs [hunk 25]
```diff
@@ -643,42 +942,83 @@ public static void Negate(ref Quaternion quaternion, out Quaternion result)
...
+		    Z *= num;
+		    W *= num;
         }
 
-
+        /// <summary>
+        /// Creates a new <see cref="Quaternion"/> that contains a normalized values from another quaternion.
+        /// </summary>
+        /// <param name="quaternion">Source <see cref="Quaternion"/>.</param>
+        /// <returns>Unit quaternion.</returns>
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```The unit length quaternion.```

File: MonoGame.Framework/Quaternion.cs [hunk 26]
```diff
@@ -643,42 +942,83 @@ public static void Negate(ref Quaternion quaternion, out Quaternion result)
...
+		    float num = 1f / ((float) Math.Sqrt((X * X) + (Y * Y) + (Z * Z) + (W * W)));
+		    X *= num;
+		    Y *= num;
+		    Z *= num;
+		    W *= num;
         }
 
-
+        /// <summary>
+        /// Creates a new <see cref="Quaternion"/> that contains a normalized values from another quaternion.
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: ```Scales the quaternion magnitude to unit length.```

File: MonoGame.Framework/Quaternion.cs (hunk 23 above)
Comment (Shqrdx) [inline MonoGame.Framework/Quaternion.cs]: The result of quaternion negation. - maybe this is better ?

File: MonoGame.Framework/Quaternion.cs [hunk 27]
```diff
@@ -776,7 +1146,11 @@ public static void Normalize(ref Quaternion quaternion, out Quaternion result)
 
         }
 
-
+        /// <summary>
+        /// Inverts values in the specified <see cref="Quaternion"/>.
+        /// </summary>
+        /// <param name="quaternion">Source <see cref="Quaternion"/> on the right of the sub sign.</param>
+        /// <returns>Result of the inversion.</returns>
         public static Quaternion operator -(Quaternion quaternion)
```
Comment (tomspilman) [inline MonoGame.Framework/Quaternion.cs]: This is negation... not inversion... those are two very different things.  Use the docs for `Negate` here.

File: MonoGame.Framework/Vector4.cs [hunk 28]
```diff
@@ -182,6 +182,18 @@ public Vector4(float value)
...
+        /// Constructs a 4d vector from <see cref="Quaternion"/>.
+        /// </summary>
+        /// <param name="quaternion">The x, y, z and w coordinates in 4d-space.</param>
+        public Vector4(Quaternion quaternion)
+        {
+            this.X = quaternion.X;
+            this.Y = quaternion.Y;
+            this.Z = quaternion.Z;
+            this.W = quaternion.W;
+        }
```
Comment (tomspilman) [inline MonoGame.Framework/Vector4.cs]: No.... lets not do this.  We have `Quaternion.ToVector4()` that is plenty.

File: MonoGame.Framework/Vector4.cs [hunk 29]
```diff
@@ -1131,6 +1143,15 @@ public override string ToString()
...
         }
 
+        /// <summary>
+        /// Gets a <see cref="Quaternion"/> representation for this object.
+        /// </summary>
+        /// <returns>A <see cref="Quaternion"/> representation for this object.</returns>
+        public Quaternion ToQuaternion()
+        {
+            return new Quaternion(this);
+        }
```
Comment (tomspilman) [inline MonoGame.Framework/Vector4.cs]: @KonajuGames - How do you feel about this?  The fact this has almost no code in it and is a very infrequent thing makes me feel it is an unnecessary "helper".

File: Test/Framework/FPointBase.cs [hunk 30]
```diff
@@ -0,0 +1,15 @@
...
+    /// <summary>
+    /// Base class for test classes with floating point operations.
+    /// </summary>
+    class FPointBase
+    {
+        protected void Compare(float expected, float source)
+        {
+            Assert.That(expected, Is.EqualTo(source).Using(FloatComparer.Epsilon));
+        }
+    }
```
Comment (tomspilman) [inline Test/Framework/FPointBase.cs]: I don't like this.  

First it seems to only be there in some effort to reduce typing... something that is not a concern.

Second it requires all tests that use this to derive from this `FPointBase` class.

File: Test/Framework/Vector2Test.cs [hunk 31]
```diff
@@ -26,53 +31,53 @@ public void Multiply()
             var vector = new Vector2(1, 2);
 
             // Test 0.0 scale.
-            Assert.AreEqual(Vector2.Zero, 0 * vector);
-            Assert.AreEqual(Vector2.Zero, vector * 0);
-            Assert.AreEqual(Vector2.Zero, Vector2.Multiply(vector, 0));
-            Assert.AreEqual(Vector2.Multiply(vector, 0), vector * 0.0f);
```
Comment (tomspilman) [inline Test/Framework/Vector2Test.cs]: These did not use an epsilon comparison on purpose... it was not a mistake.

Comment (Shqrdx) [inline Test/Framework/Vector2Test.cs]: Assert.AreEqual for floats need to be reverted too ?

**Comment by tomspilman** [comment]
@Shqrdx - So I think I've spent maybe 2 hours now on this PR and still feel it needs a bunch more work beyond what I just covered.

@KonajuGames and I need to rethink how documentation tasks are handled moving forward.  I suggest you don't submit any new docs until then.

File: Test/Framework/Vector2Test.cs (hunk 31 above)
Comment (tomspilman) [inline Test/Framework/Vector2Test.cs]: Yes... there was nothing wrong with all the existing tests.  They all passed.

This is also the sort of thing that should have all gone into another PR.  Lumping it in here with a ton of doc changes just complicated things further.

**Comment by Shqrdx** [comment]
Hm, I think I push only the Quaternion docs and it will be fine. I will do it in new PR.

**Comment by tomspilman** [comment]
> Hm, I think I push only the Quaternion docs and it will be fine

Great!

**Pull request closed without merging**